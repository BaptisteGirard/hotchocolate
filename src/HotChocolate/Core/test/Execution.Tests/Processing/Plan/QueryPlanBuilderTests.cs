using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using HotChocolate.Language;
using HotChocolate.StarWars;
using Snapshooter;
using Snapshooter.Xunit;
using Xunit;

namespace HotChocolate.Execution.Processing.Plan
{
    public class QueryPlanBuilderTests
    {
        [Fact]
        public void GetHero_Plan()
        {
            // arrange
            ISchema schema = SchemaBuilder.New()
                .AddStarWarsTypes()
                .Create();

            DocumentNode document = Utf8GraphQLParser.Parse(
                @"query GetHero($episode: Episode, $withFriends: Boolean!) {
                    hero(episode: $episode) {
                        name
                        friends @include(if: $withFriends) {
                            nodes {
                                id
                            }
                        }
                    }
                }");

            OperationDefinitionNode operationDefinition =
                document.Definitions.OfType<OperationDefinitionNode>().Single();

            IPreparedOperation operation =
                OperationCompiler.Compile(
                    "a",
                    document,
                    operationDefinition,
                    schema,
                    schema.QueryType);

            // act
            QueryPlanNode root = QueryPlanBuilder.BuildNode(operation);

            // assert
            
            Snapshot(operation, root);
        }

        private static void Snapshot(IPreparedOperation operation, QueryPlanNode node)
        {
            using var stream = new MemoryStream();
            using var writer = new Utf8JsonWriter(stream, new JsonWriterOptions { Indented = true});
            node.Serialize(writer);
            writer.Flush();
            Encoding.UTF8.GetString(stream.ToArray())
                .MatchSnapshot(new SnapshotNameExtension("plan"));
            operation.Print().MatchSnapshot(new SnapshotNameExtension("operation"));
        }
    }
}
