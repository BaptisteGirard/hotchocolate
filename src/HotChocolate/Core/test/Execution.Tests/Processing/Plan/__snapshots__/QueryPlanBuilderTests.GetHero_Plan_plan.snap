{
  "type": "Parallel",
  "nodes": [
    {
      "type": "Sequence",
      "nodes": [
        {
          "type": "Resolver",
          "strategy": "Parallel",
          "selections": [
            {
              "id": 0,
              "field": "Query.hero",
              "responseName": "hero"
            },
            {
              "id": 2,
              "field": "Human.friends",
              "responseName": "friends"
            },
            {
              "id": 7,
              "field": "Droid.friends",
              "responseName": "friends"
            }
          ]
        },
        {
          "type": "Resolver",
          "strategy": "Serial",
          "selections": [
            {
              "id": 1,
              "field": "Human.name",
              "responseName": "name"
            }
          ]
        }
      ]
    },
    {
      "type": "Resolver",
      "strategy": "Parallel",
      "selections": [
        {
          "id": 3,
          "field": "CharacterConnection.nodes",
          "responseName": "nodes",
          "pure": true
        },
        {
          "id": 8,
          "field": "CharacterConnection.nodes",
          "responseName": "nodes",
          "pure": true
        }
      ]
    }
  ]
}
